File: tools/c-api-jsc-diff.py
#!/usr/bin/env python3
"""Compile the value/class API fixture against system JSC and compare output."""

from __future__ import annotations

import difflib
import hashlib
import platform
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
FIXTURE = ROOT / "tests/c_api_value_diff.c"


def run(command: list[str]) -> bytes:
    return subprocess.run(command, cwd=ROOT, check=True, stdout=subprocess.PIPE).stdout


def main() -> None:
    if len(sys.argv) != 2:
        raise SystemExit("usage: c-api-jsc-diff.py <zig-js fixture executable>")
    if platform.system() != "Darwin":
        raise SystemExit("the pinned JavaScriptCore differential gate requires macOS")

    sdk_root = run(["xcrun", "--sdk", "macosx", "--show-sdk-path"]).decode().strip()
    subprocess.run(
        [sys.executable, "tools/verify-c-api.py", "--sdk-root", sdk_root],
        cwd=ROOT,
        check=True,
    )
    with tempfile.TemporaryDirectory(prefix="zig-js-jsc-diff-") as directory:
        reference = Path(directory) / "jsc-value-diff"
        subprocess.run(
            [
                "xcrun", "--sdk", "macosx", "clang",
                str(FIXTURE), "-framework", "JavaScriptCore", "-o", str(reference),
            ],
            cwd=ROOT,
            check=True,
        )
        expected = run([str(reference)])
        actual = run([sys.argv[1]])

    if actual != expected:
        difference = difflib.unified_diff(
            expected.decode().splitlines(),
            actual.decode().splitlines(),
            fromfile="system JavaScriptCore",
            tofile="zig-js",
            lineterm="",
        )
        raise SystemExit("C API differential mismatch:\n" + "\n".join(difference))
    digest = hashlib.sha256(actual).hexdigest()[:16]
    print(f"c-api JSC differential: matched {len(actual.splitlines())} rows ({digest})")


if __name__ == "__main__":
    main()

File: tools/c-api-jsc-diff.ts
/** Compile C API fixtures against system JSC and compare their output. */
import { checked, removeTemporaryDirectory, sha256File, temporaryDirectory, writeText, run } from "./lib/home";
declare const __dirname: string;
declare function require(specifier: string): any;
const { verifyCAPI } = require("./verify-c-api");
// The imported verifier is the executable gate at tools/verify-c-api.ts.
const ROOT = __dirname === "tools" ? "." : __dirname.slice(0, __dirname.lastIndexOf("/tools"));
const join = (left: string, right: string): string => `${left.replace(/\/$/, "")}/${right}`;
const fixtures = [join(ROOT, "tests/c_api_value_diff.c"), join(ROOT, "tests/c_api_context_group_diff.c")], args = process.argv.slice(2);
if (args.length !== fixtures.length) throw new Error("usage: c-api-jsc-diff.ts <zig-js value fixture> <zig-js context-group fixture>");
if (checked(["uname", "-s"], "detect platform").trim() !== "Darwin") throw new Error("the pinned JavaScriptCore differential gate requires macOS");
const sdkRoot = checked(["xcrun", "--sdk", "macosx", "--show-sdk-path"], "locate macOS SDK").trim();
verifyCAPI(sdkRoot);
const directory = temporaryDirectory("zig-js-jsc-diff"), expectedParts: string[] = [], actualParts: string[] = [];
try {
  fixtures.forEach((fixture, index) => {
    const reference = join(directory, `jsc-diff-${index}`);
    checked(["xcrun", "--sdk", "macosx", "clang", fixture, "-framework", "JavaScriptCore", "-o", reference], `compile system JSC fixture ${fixture}`);
    expectedParts.push(checked([reference], `run ${reference}`));
    actualParts.push(checked([args[index]], `run ${args[index]}`));
  });
  const expected = expectedParts.join(""), actual = actualParts.join("");
  if (actual !== expected) {
    const expectedPath = join(directory, "system-jsc.txt"), actualPath = join(directory, "zig-js.txt");
    writeText(expectedPath, expected); writeText(actualPath, actual);
    const difference = run(["diff", "-u", expectedPath, actualPath]);
    throw new Error(`C API differential mismatch:\n${difference.stdout || difference.stderr}`);
  }
  const output = join(directory, "actual.txt"); writeText(output, actual);
  const rows = actual ? actual.replace(/\n$/, "").split("\n").length : 0;
  console.log(`c-api JSC differential: matched ${rows} rows (${sha256File(output).slice(0, 16)})`);
} finally { removeTemporaryDirectory(directory); }
